File: frontend/src/java/com/agnitas/util/PubID.java
/*

    Copyright (C) 2025 AGNITAS AG (https://www.agnitas.org)

    This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
    You should have received a copy of the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.

*/

package com.agnitas.util;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * This class provides the generation and decoding of an agnPUBID,
 * an ID like the agnUID, but used for public accessable information.
 * At this state, this is used to access an anonymous fullview as
 * posted in a social network. The ID provdies the information of
 * the recipient of whom the data should be accessed and is not
 * intended to be a secure piece of information.
 * 
 * The pubID transports the mailingID of the related mailing and
 * the customerID of the recipient. An optional source (free text)
 * to indicate from where the ID is used (e.g. a short name of the
 * social network) and on optional parameter which can be used by
 * the code which processes this ID and depends on the context it
 * is used in.
 */
public class PubID {
	private final static String	cl = "w5KMCHOXE_PTuLcfF6D1ZI3BydeplQaztVAnUj0bqos7k49YgWhxiS-RrGJm8N2v";
	private final static Pattern	srcInvalid = Pattern.compile ("[^0-9a-zA-Z_-]");
	private long			mailingID;
	private long			customerID;
	private String			source;
	private String			parm;
	
	public PubID () {
		mailingID = -1;
		customerID = -1;
		source = null;
		parm = null;
	}
	
	public long getMailingID () {
		return mailingID;
	}
	public void setMailingID (long nMailingID) {
		mailingID = nMailingID;
	}
	public long getCustomerID () {
		return customerID;
	}
	public void setCustomerID (long nCustomerID) {
		customerID = nCustomerID;
	}
	public String getSource () {
		return source;
	}
	public void setSource (String nSource) {
		if (nSource == null) {
			source = nSource;
		} else {
			Matcher	m = srcInvalid.matcher (nSource);
		
			if (! m.find ()) {
				source = nSource;
			} else {
				StringBuffer	scratch = new StringBuffer (nSource.length ());
				int		last, pos;
			
				last = 0;
				do {
					pos = m.start ();
					if (last < pos)
						scratch.append (nSource.substring (last, pos));
					scratch.append ('_');
					last = ++pos;
				}	while (m.find (pos));
				if (pos < nSource.length ())
					scratch.append (nSource.substring (pos));
				source = scratch.toString ();
			}
			if ((source != null) && (source.length () > 20)) {
				source = source.substring (0, 20);
			}
		}
	}
	public String getParm () {
		return parm;
	}
	public void setParm (String nParm) {
		parm = nParm;
	}

	public String createID () {
		String	src = mailingID + ";" + customerID + ";" +
			      (source == null ? "" : source) + (parm == null || parm.length () == 0 ? "" : ";" + parm);
		return encode (src);
	}
	
	public String createID (long cMailingID, long cCustomerID, String cSource, String cParm) {
		setMailingID (cMailingID);
		setCustomerID (cCustomerID);
		setSource (cSource);
		setParm (cParm);
		return createID ();
	}

	public String createID (long cMailingID, long cCustomerID, String cSource) {
		return createID (cMailingID, cCustomerID, cSource, null);
	}

	/**
	 * parses a uid and populate its instance variables
	 * with the content of the ID, if parsing had been
	 * successful.
	 * 
	 * @param pid the pubID to be parsed
	 * @return    true, if parsing had been successful, false otherwise
	 */
	public boolean parseID (String pid) {
		boolean	rc;
		String	dst = decode (pid);

		rc = false;
		if (dst != null) {
			String[]	parts = dst.split (";", 4);
		
			if ((parts.length == 3) || (parts.length == 4)) {
				long	tMailingID;
				long	tCustomerID;
				String	tSource;
				String	tParm;
				
				try {
					tMailingID = Long.parseLong (parts[0]);
					tCustomerID = Long.parseLong (parts[1]);
					tSource = parts[2];
					tParm = parts.length == 4 ? parts[3] : null;
					if ((tMailingID > 0) && (tCustomerID > 0)) {
						setMailingID (tMailingID);
						setCustomerID (tCustomerID);
						setSource (tSource);
						setParm (tParm);
						rc = true;
					}
				} catch (NumberFormatException e) {
					// do nothing
				}
			}
		}
		return rc;
	}

	private char checksum (String s) {
		int	cs;
		int	slen;
		
		cs = 12;
		slen = s.length ();
		for (int n = 0; n < slen; ++n) {
			cs += s.charAt (n);
			
		}
		return cl.charAt (cs & 0x3f);
	}
	
	private String encode (String s) {
		String	rc;
		
		try {
			byte[]		dump = s.getBytes ("UTF8");
			StringBuffer	temp = new StringBuffer (((dump.length + 3) * 4) / 3);

			for (int n = 0; n < dump.length; n += 3) {
				int	d = ((dump[n] & 0xff) << 16) |
					    ((n + 1 < dump.length ? dump[n + 1] & 0xff : 0) << 8) |
					    (n + 2 < dump.length ? dump[n + 2] & 0xff : 0);
				int	c1 = (d >> 18) & 0x3f,
					c2 = (d >> 12) & 0x3f,
					c3 = (d >> 6) & 0x3f,
					c4 = d & 0x3f;
				temp.append (cl.charAt (c1));
				temp.append (cl.charAt (c2));
				temp.append (cl.charAt (c3));
				temp.append (cl.charAt (c4));
			}
			temp.insert (5, checksum (temp.toString ()));
			rc = temp.toString ();
		} catch (java.io.UnsupportedEncodingException e) {
			rc = null;
		}
		return rc;
	}
	
	private String decode (String s) {
		String	rc;
		int	slen;
		
		rc = null;
		slen = s.length ();
		if ((slen > 5) && (((slen - 1) & 3) == 0)) {
			boolean	st = true;
			char	check = s.charAt (5);
			byte[]	collect;
			int	pos;
			
			s = s.substring (0, 5) + s.substring (6);
			if (check == checksum (s)) {
				--slen;
				collect = new byte[slen];
				pos = 0;
				for (int n = 0; st && (n < slen); n += 4) {
					int	v1 = cl.indexOf (s.charAt (n)),
						v2 = cl.indexOf (s.charAt (n + 1)),
						v3 = cl.indexOf (s.charAt (n + 2)),
						v4 = cl.indexOf (s.charAt (n + 3));

					if ((v1 == -1) || (v2 == -1) || (v3 == -1) || (v4 == -1))
						st = false;
					else {
						int	v = (v1 << 18) | (v2 << 12) | (v3 << 6) | v4;

						collect[pos++] = (byte) ((v >> 16) & 0xff);
						collect[pos++] = (byte) ((v >> 8) & 0xff);
						collect[pos++] = (byte) (v & 0xff);
					}
				}
				if (st) {
					while ((pos > 0) && (collect[pos - 1] == 0))
						--pos;
					try {
						rc = new String (collect, 0, pos, "UTF8");
					} catch (java.io.UnsupportedEncodingException e) {
						rc = null;
					}
				}
			}
		}
		return rc;
	}
}

File: backend/src/script/lib/agn3/pubid.py
####################################################################################################################################################################################################################################################################
#                                                                                                                                                                                                                                                                  #
#                                                                                                                                                                                                                                                                  #
#        Copyright (C) 2025 AGNITAS AG (https://www.agnitas.org)                                                                                                                                                                                                   #
#                                                                                                                                                                                                                                                                  #
#        This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.    #
#        This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.           #
#        You should have received a copy of the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.                                                                                                            #
#                                                                                                                                                                                                                                                                  #
####################################################################################################################################################################################################################################################################
#
from	__future__ import annotations
import	re
from	typing import Optional, Union
from	typing import List, Sequence
from	agn3.ignore import Ignore
from	agn3.stream import Stream
#
__all__ = ['PubID']
#
class PubID:
	"""Handled public id agnPUBID

this handles the public id agnPUBID which is used for the anon
preview."""
	__slots__ = ['mailing_id', 'customer_id', 'source', 'selector']
	scramble = 'w5KMCHOXE_PTuLcfF6D1ZI3BydeplQaztVAnUj0bqos7k49YgWhxiS-RrGJm8N2v'
	source_invalid = re.compile ('[^0-9a-zA-Z_-]')

	def __init__ (self) -> None:
		self.mailing_id = 0
		self.customer_id = 0
		self.source: Optional[str] = None
		self.selector: Optional[str] = None
	
	def __str__ (self) -> str:
		return '{name} <{parameter}>'.format (
			name = self.__class__.__name__,
			parameter = (Stream ([
					('mailing', self.mailing_id),
					('customer', self.customer_id),
					('source', self.source),
					('selector', self.selector)
				])
				.filter (lambda kv: kv[1] is not None)
				.map (lambda kv: '{key}={value!r}'.format (key = kv[0], value = kv[1]))
				.join (', ')
			)
		)

	def __source (self, source: Optional[str]) -> Optional[str]:
		if source is not None:
			if len (source) > 20:
				source = source[:20]
			source = self.source_invalid.subn ('_', source)[0]
		return source

	def __checksum (self, s: Sequence[str]) -> str:
		cs = 12
		for ch in s:
			cs += ord (ch)
		return self.scramble[cs & 0x3f]

	def __encode (self, source: Union[str, bytes]) -> str:
		s: bytes = source.encode ('UTF-8') if isinstance (source, str) else source
		slen = len (s)
		temp: List[str] = []
		n = 0
		while n < slen:
			chk = s[n:n + 3]
			d = chk[0] << 16
			if len (chk) > 1:
				d |= chk[1] << 8
				if len (chk) > 2:
					d |= chk[2]
			n += 3
			temp.append (self.scramble[(d >> 18) & 0x3f])
			temp.append (self.scramble[(d >> 12) & 0x3f])
			temp.append (self.scramble[(d >> 6) & 0x3f])
			temp.append (self.scramble[d & 0x3f])
		temp.insert (5, self.__checksum (temp))
		return ''.join (temp)

	def __decode (self, s: str) -> Optional[str]:
		rc: Optional[str] = None
		slen = len (s)
		if slen > 5 and (slen - 1) & 3 == 0:
			check = s[5]
			s = s[:5] + s[6:]
			if check == self.__checksum (s):
				slen -= 1
				collect = []
				ok = True
				n = 0
				while n < slen and ok:
					v = [self.scramble.find (s[_c]) for _c in range (n, n + 4)]
					n += 4
					if -1 in v:
						ok = False
					else:
						d = (v[0] << 18) | (v[1] << 12) | (v[2] << 6) | v[3]
						collect.append (chr ((d >> 16) & 0xff))
						collect.append (chr ((d >> 8) & 0xff))
						collect.append (chr (d & 0xff))
				if ok:
					while collect and collect[-1] == '\0':
						collect = collect[:-1]
					rc = ''.join (collect)
		return rc

	def create (self, mailing_id: Optional[int] = None, customer_id: Optional[int] = None, source: Optional[str] = None, selector: Optional[str] = None) -> str:
		"""Creates a new agnPUBID

creates a new public id which can be used for the anon preview. Either
set the variables here or set them by assigning the corrosponding
instance variables."""
		if mailing_id is None:
			mailing_id = self.mailing_id
		if customer_id is None:
			customer_id = self.customer_id
		if source is None:
			source = self.source
			if source is None:
				source = ''
		if selector is None:
			selector = self.selector
			if not selector:
				selector = None
		src = '{mailing_id};{customer_id};{source}'.format (
			mailing_id = mailing_id,
			customer_id = customer_id,
			source = self.__source (source)
		)
		if selector is not None:
			src += f';{selector}'
		return self.__encode (src)

	def parse (self, pid: str) -> bool:
		"""Parses an agnPUBID
		
if parsing had been successful, the instance variable containing the
parsed content and the method returns True, otherwise the content of
the instance variables is undefinied and the method returns False."""
		rc = False
		dst = self.__decode (pid)
		if dst is not None:
			parts = dst.split (';', 3)
			if len (parts) in (3, 4):
				with Ignore (ValueError):
					mailing_id = int (parts[0])
					customer_id = int (parts[1])
					source = parts[2]
					if len (parts) > 3:
						selector: Optional[str] = parts[3]
					else:
						selector = None
					if mailing_id > 0 and customer_id > 0:
						self.mailing_id = mailing_id
						self.customer_id = customer_id
						if source:
							self.source = source
						else:
							self.source = None
						self.selector = selector
						rc = True
		return rc

